### ISSUE TITLE
[BUG] Cpu usage extremely high when stdout=/dev/null but stdin isn't

### ISSUE THREAD

**Issue by Ae3NerdGod** [BeamMP/BeamMP-Server#19]
So the cpu usage on my i3 when using the release (2.0.3) to run more than one server is 100%, but after I've manually compiled the v2.0.3 release, the cpu just idles as if the (5+) beamMP servers aren't even running.

Server works fine in both instances, except that when using the release it overheats my i3 and eventually will shut it down.

**Comment by lionkor** [comment]
Hello!

Please, which OS and version are you on, and whta is the exact CPU you're running?

Each server uses multiple cores, so with many servers it will use a lot of the CPU.

**Comment by Ae3NerdGod** [comment]
https://github.com/BeamMP/BeamMP-Server/issues/17#issuecomment-816973135

**Comment by Ae3NerdGod** [comment]
have you been able to look more into this? im currently running 6 servers on said i3 and it keeps 3 of the cores at 100% at all times, is that normal? 

including info from that other post to be through:
[ldd_vers.txt](https://github.com/BeamMP/BeamMP-Server/files/6359525/ldd_vers.txt)
[lsb_release.txt](https://github.com/BeamMP/BeamMP-Server/files/6359526/lsb_release.txt)
[lscpu.txt](https://github.com/BeamMP/BeamMP-Server/files/6359527/lscpu.txt)
[uname.txt](https://github.com/BeamMP/BeamMP-Server/files/6359528/uname.txt)

**Comment by Ae3NerdGod** [comment]
also are all these extra processes normal?

![Screenshot 2021-04-22 114154](https://user-images.githubusercontent.com/2036505/115743585-cd81f980-a35f-11eb-95cd-cef1b2d824bf.png)

**Comment by lionkor** [comment]
Hello,

in `htop` you can press Shift+H to hide userland threads.

Those "processes" you're seeing are all supposed to be there, yeah. All the green ones are threads, so they are pieces of code that are executing in parallel. This is used to make it possible to handle multiple players at the same time, so those are needed, yes. 

As you can see, most of them are around 0-1% CPU usage.

**Comment by lionkor** [comment]
Okay, so I can't seem to reproduce this myself.

Of course a little performance increase is expected when you compile it yourself, but it should never *ever* be this high. 

I'd consider this a bug on our side, but I need to be able to reproduce it, and currently I don't seem to be able to. Could you possibly install `perf` and run:

`perf record ./BeamMP-Server-linux`

The issue with the 100% CPU spin needs to happen now. Once it happens, type `exit()` (with parentheses) in the server's commandline and wait for everything to terminate. Afterwards, you should see a file called `perf.data` in the same directory. 

Please attach this file here. This will record a performance profile and we'll be able to see where the 100% cpu spin is happening.

**Comment by lionkor** [comment]
Any update on this, @Ae3NerdGod?

**Comment by Ae3NerdGod** [comment]
Sorry, when I tried to install perf, it busted my linux Install. I think all the problems I'm having are stemming from the cpu I was using. Were getting the parts together for a ryzen rig, if I'm still having the problem I'll re-open

**Comment by Ae3NerdGod** [comment]
ive finally migrated the servers to a new server and i figured out the 100% CPU usage issue. it occurs when using a script, to start a script, that starts beammp-server. ill explain a bit further. 

 i make a script that creates the server cfg based on variables in the script and then cd's to a specific directory via variable, bind mounts some common mod zips to the Client folder then starts the beam server in that directory. this allows me to make a new server quickly by copying and editing one file. -- this all works perfectly when started via the script itself ie `sh ./roane.svr` 
 
 but when i use a script that `for` loops through all the .svr files i have and starts them, then exits, it causes the 100% cpu problem. for now ive mitigated this problem by making separate .service files for each server and instead starting them via `systemctl` using the `for` loop

TL;DR : using a script, to start multiple scripts that start beam servers, causes the 100% cpu problem.

**Comment by Starystars67** [comment]
I wonder if this is because the service (script) that hosted the initial loop finds that there is no running process once the loop finishes and as a result starts its self once again (Starting all of the servers again) and as this happens again and again it forces the CPU usage and logically RAM resources up too.

Equally I could have misunderstood what you are saying.

**Comment by lionkor** [comment]
So this is (was) a systemd issue. Each systemd service is supposed to always run one process, not start multiple. Anyone else running into issues starting a BeamMP server via systemd should read [this short guide](https://www.shubhamdipt.com/blog/how-to-create-a-systemd-service-in-linux/).

**Comment by Ae3NerdGod** [comment]
> So this is (was) a systemd issue. Each systemd service is supposed to always run one process, not start multiple. Anyone else running into issues starting a BeamMP server via systemd should read [this short guide](https://www.shubhamdipt.com/blog/how-to-create-a-systemd-service-in-linux/). 
> 
> 

Well, I used systemd to fix it (by running each one individually), starting my original script from terminal caused the issue as well, and I used mate-desktops "startup applications" to start the script previously and was still getting this issue. My first time using systemd to start the scripts were what fixed my issue, so I don't think systemd was causing it.

**Comment by Ae3NerdGod** [comment]
> I wonder if this is because the service (script) that hosted the initial loop finds that there is no running process once the loop finishes and as a result starts its self once again (Starting all of the servers again) and as this happens again and again it forces the CPU usage and logically RAM resources up too.
> 
> Equally I could have misunderstood what you are saying.

I triple checked for that, and even made a script with no loop (just starts each server-starter script and exits) still resulted in high CPU. It should be reproducible

**Comment by lionkor** [comment]
That makes sense, reopening.

**Comment by Ae3NerdGod** [comment]
> That makes sense, reopening.

For me it is reproducible with the following steps

Make 2 scrips, parent.sh, child.sh, in the same dir as the beam server executable.

In parent.sh;
`child.sh &`

In child.sh;
`Beam-Server-Latest`

This should cause one core to be at 90-100%

2 servers started via ampersand: ```Beam-Server-Latest &
Beam-Server-Latest
```
Should cause 2 cores to be affected.

If you add `sleep 5` to the end of `parent.sh` the affect should be delayed 5 seconds

**Comment by edrcq** [comment]
The reason is "stdin" is listening in a loop... I think.

I started the program by sending EOF to "stop" the stdin listening to fix it.

stdin is the console on the server where you can type once the server is started.

When I close the server, the stdin is always listening, like if the server is not closed. (but it is, no process running)

Probably a low level bug on the standard inpup ?

**Comment by Ae3NerdGod** [comment]
The stdin loop might explain why It (used to) occasionally make my proxmox console or putty "[divide by zero](https://giphy.com/gifs/reaction-mood-1k03CFEqbh0tneLeZi)"

If I Ctrl+c'd the script It would exit but then repeat the prompt like "root@beamsvr:~# root@beamsvr:~#root@beamsvr:~#root@beamsvr:~#root@beamsvr:~#...."
Until I forced the container to stop (proxmox) or killed the terminal window (bare metal)

Also, occasionally it would take me back to a prompt, but terminal didn't process line breaks correctly, so if I pressed enter another prompt would show up after the command on the same line.

Idk if any of this info helps, but it sounds related

**Comment by lionkor** [comment]
That is strange, I thought https://github.com/lionkor/commandline had that fixed... Well, in that case, I'll look into that issue again.

For anyone wanting to help - the stdin loop happens when stdin is /dev/null, and usually you can detect this by asking if stdin is a tty (`isatty(stdin)` or something like that).

**Comment by lionkor** [comment]
I am not able to reproduce this. Could you tell me which version of the server you are running? @Ae3NerdGod 
Just dump your `git show` here.

**Comment by lionkor** [comment]
Can you please verify that this still happens with the latest server (v2.1.1 or v2.1.2)?

**Comment by lionkor** [comment]
okay i could reproduce this (by accident) with `screen -dmS beam ./bin/BeamMP-Server`.

**Comment by synth80s** [comment]
> okay i could reproduce this (by accident) with `screen -dmS beam ./bin/BeamMP-Server`.

Hey lionkor, I am experiencing a similar issue. I have a simple script that starts the server simply cd to the directory and ./Beammp_server

This causes the threads in my virtual machine to peg at 100%. If I start the server manually from the command line this does not happen. I didn't notice this issue until the update.

**Comment by kevinsynnott** [comment]
I was having the same issues with a fresh install on Ubuntu 20.04. No issues when running direct from the command line. I tried running with cron @reboot and also a a systemd service and both times, had issues with ram usuage climbing until full and also cpu usuage high.

I have managed to get it to work as systemd service and seems to be running normally as if I ran from the command line.

```
[Unit]
Description=BeamMP Server Service
After=getty.service

[Service]
ExecStart=/home/kevin/BeamMP-Server-linux
WorkingDirectory=/home/kevin
StandardInput=tty-force
TTYVHangup=yes
TTYPath=/dev/tty20
TTYReset=yes
Type=simple
RemainAfterExit=false
Restart=always
RestartSec=5s

[Install]
WantedBy=multi-user.target
```
 
I found an example [https://unix.stackexchange.com/questions/453998/systemd-connect-to-stdin-stdout-after-service-has-started](url)

I am not advanced enough to expain what this is doing, I just figured it out from a bit of searching from what lionkor was explaining about stdin. Someone else might be better able to understand or even improve on this way.

**Comment by Ae3NerdGod** [comment]
i have a workaround with a bit of bash, it basically keeps restarting the server until it starts correctly:

```
function StartBeam(){
maptmp=$(cat ServerConfig.toml | grep -i map | awk '{print $3}' | sed s/\'// | sed s/\'// | sed -e 's/\/levels\///' | sed -e 's/\/info.json//')

killsize=10
outfile=$outfile_$maptmp
rm -f $outfile
sleep 1
sync
nohup ./BeamMP-Server-linux > $outfile 2>&1 &
pid=$!
 
trap "kill -9 $pid 2> /dev/null" EXIT

while kill -0 $pid 2> /dev/null; do
	i=$((i+1))
    sleep 1
	size=$(du $outfile | awk '{print $1}' )
	echo output size is $size
	if [ $size -gt $killsize ]
	then
		echo Hey that\'s a large number. killing $pid
		kill -9 $pid > /dev/null 2>&1
		sleep 2
		StartBeam
		break
	fi
	if [ $i -gt 10 ]
	then
		echo started beam correctly
		while kill -0 $pid 2> /dev/null; do
                        #keep script open while beam is open
			sleep 30
		done
	fi
done

# Disable the trap on a normal exit.
trap - EXIT
}
```

**Comment by lionkor** [comment]
Nice! I want to quickly say that parsing toml like that does not follow the toml spec, so be aware to limit your config and check it after every update, as the server may modify it at any time according to the TOML spec.

Nice workaround. I'm still trying to find the core issue.

**Comment by Ae3NerdGod** [comment]
> Nice! I want to quickly say that parsing toml like that does not follow the toml spec, so be aware to limit your config and check it after every update, as the server may modify it at any time according to the TOML spec.
> 
> Nice workaround. I'm still trying to find the core issue.

im completely unaware of the intricacies of toml, feel free to improve lol

and no worries, got it working for now.

**Comment by Ae3NerdGod** [comment]
> im completely unaware of the intricacies of toml, feel free to improve lol

i just modified it to not require the toml at all, and use a random file name instead (and clean it up)

its only purpose was to allow for multiple servers starting at the same time

**Comment by lionkor** [comment]
The workaround until the fix is in the stable release is to set stdin to /dev/null

**Comment by Ae3NerdGod** [comment]
> The workaround until the fix is in the stable release is to set stdin to /dev/null


for the record, ive seen it go to 100% using `> /dev/null 2>&1 ` to redirect stderr and stdin. this was, again, while running in a script that starts more than one beam server. but it is far less likely from my experience. 

perhaps it was nohup causing it in that situation now that i think of it.

**Comment by 20dka** [comment]
I was having high CPU usage on Linux when running with a systemd service, I fixed it by adding `StandardInput=null` to the service config

**Comment by lionkor** [comment]
> using `> /dev/null 2>&1` to redirect stderr and stdin

`>` is the same as `1>`, 1 is `stdout`, `2>` is 2 which is stderr. This is the same as `&>` in common shells.

You're redirecting std**out** and std**err**, but std**in** is still there. That's why it still breaks.

The way to pipe stdin, stdout and stderr to /dev/null is `./BeamMP-Server-linux </dev/null &>/dev/null`. 
Here `<` is stdin.

This is obviously not optimal, and in the next release piping stdout to /dev/null is going to be enough. For now, redirect stdin  to /dev/null.

**Comment by lionkor** [comment]
For the record, I wasn't aware of this either. The BeamMP-Server correctly turns off the input handling when stdin is /dev/null, but rarely anyone sets that - I thought that setting stdout to /dev/null also affected stdin in some way - this obviously isn't true and wouldn't make sense. 

This is why it spins at 100%, since it tries to read but stdin is just dangling in a screen, reading an endless stream of zeroes. 

As I said, fixed in the next update.

**Comment by Ae3NerdGod** [comment]
> You're redirecting std**out** and std**err**, but std**in** is still there. That's why it still breaks.

ahhhhhh, i read your stdin as stdout for some reason. 

tested `</dev/null` and it seems to work, much cleaner workaround.

**Comment by Ae3NerdGod** [comment]
> This is why it spins at 100%, since it tries to read but stdin is just dangling in a screen, reading an endless stream of zeroes.

this makes so much sense now, 

almost like it was trying to read `screen`, `nohup` or putty's terminal multiplexer as stdin? (for me the issue never occurred on a terminal window or Ctrl+Alt+F# no matter how hard i tried)

**Comment by lionkor** [comment]
> almost like it was trying to read screen, nohup or putty's terminal multiplexer as stdin

I guess so, yeah! It had a valid `stdin`, but and repeatedly read what I assume were zero-bytes. 

Glad that this workaround worked for you, public version fix coming with the next update (that sets stdin to /dev/null if stdout is /dev/null, basically)

**Comment by lionkor** [comment]
There seems to still be an instance of this happening, looks like possibly some spinning caused by a race condition somewhere. @20dka and I both encountered this on multiple instances, usually when starting a BeamMP-Server as a systemd service (and setting stdin & stdout to /dev/null).

**Comment by Ae3NerdGod** [comment]
i wonder if it could be reading its own stdout from stdin? try setting both to null, it hasnt happend to me once since i set stdin as well as stdout and i just use the logfile for console output via OpenGamePanel

**Comment by Ae3NerdGod** [comment]
two things: 
1) it happens when using nohup toredirect nohup, not just when stdout is /dev/null 

> > but repeatedly read what I assume were zero-bytes.
2) i guess i neglected to mention because i thought you had it figured out, but if you use nohup to capture the output when it happens, its definitely not zero bytes.
